colossalai/inference/tensor_parallel/modeling/llama.py

```diff
@@ -348,35 +385,27 @@ def llama_flash_attn_kvcache_forward(
                     infer_state.decode_mem_index,
                     infer_state.cache_manager,
                 )
-
-            # second token and follows
-            # kv = torch.stack((key_states, value_states), dim=2)
-            # (batch_size, seqlen, nheads, headdim)
-            attn_output = torch.empty_like(query_states)
-
-            if self.num_key_value_groups == 1:
-                token_attention_fwd(
-                    query_states,
-                    infer_state.cache_manager.key_buffer[infer_state.decode_layer_id],
-                    infer_state.cache_manager.value_buffer[infer_state.decode_layer_id],
-                    attn_output,
-                    infer_state.block_loc,
-                    infer_state.start_loc,
-                    infer_state.seq_len,
-                    infer_state.cache_manager.past_key_values_length,
-                )
+            
+            if HAS_LIGHTLLM_KERNEL:
+                attn_output = torch.empty_like(query_states)
+                llama_triton_token_attention(query_states, attn_output, infer_state, num_key_value_groups=self.num_key_value_groups)
             else:
-                Llama2TokenAttentionForwards.token_attn(
-                    query_states,
-                    infer_state.cache_manager.key_buffer[infer_state.decode_layer_id],
-                    infer_state.cache_manager.value_buffer[infer_state.decode_layer_id],
-                    attn_output,
-                    infer_state.block_loc,
-                    infer_state.start_loc,
-                    infer_state.seq_len,
-                    infer_state.cache_manager.past_key_values_length,
-                    infer_state.other_kv_index,
-                )
+                heads_per_group = self.num_heads // self.num_key_value_heads
+                cache_k = infer_state.cache_manager.key_buffer[infer_state.decode_layer_id]
+                cache_v = infer_state.cache_manager.value_buffer[infer_state.decode_layer_id]
+                
+                query_states = query_states.view(bsz, -1, self.num_heads, self.head_dim)
+                copy_cache_k= copy.deepcopy(cache_k)
+                copy_cache_v = copy.deepcopy(cache_v)
+                copy_cache_k = copy_cache_k.view(bsz, -1, self.num_key_value_heads, self.head_dim)
+                copy_cache_v = copy_cache_v.view(bsz, -1, self.num_key_value_heads, self.head_dim)
```

Comment:
`copied_kv` need to be delete after flash decoding?

Reply:
fixed.